designs/node-overlays.md

```diff
@@ -0,0 +1,186 @@
+
+## Use Cases
+
+### Price
+
+Each instance offering comes with a price, which is used as a minimization function in Karpenter's scheduling algorithm. However, Karpenter cannot perfectly understand all factors that influence price, such as an external vendor's fee, a pricing discount, or an external motiviation such as carbon-offset.
+
+* https://github.com/aws/karpenter-provider-aws/pull/4686
+* https://github.com/aws/karpenter-provider-aws/issues/3860
+* https://github.com/aws/karpenter-provider-aws/pull/4697
+
+### Extended Resources
+
+InstanceTypes have a set of well known resource capacities, such as CPU, memory, or GPUs. However, Kubernetes supports arbitrary [extended resources](https://kubernetes.io/docs/tasks/administer-cluster/extended-resource-node/), which cannot be known prior to Node launch. Karpenter lacks a mechanism to teach the simulation about extended resource capacities.
+
+* https://github.com/kubernetes-sigs/karpenter/issues/751
+* https://github.com/kubernetes-sigs/karpenter/issues/729
+
+### Resource Overhead
+
+* https://github.com/aws/karpenter-provider-aws/issues/5161
+
+System software such as containerd, kubelet, and the host operating system come with resource overhead like CPU and memory, which is substracted from a Node's capacity. Karpenter attempts to calculate this overhead where it is known, but cannot predict the overhead required by custom operating systems. To complicate this, the overhead often changes for each version of an operating system.
+
+## Proposal
+
+NodeOverlays are a new Karpenter API that enable customers to fine-tune the scheduling simulation for advanced use cases. NodeOverlays can be configured with a `metav1.Selector` which flexibly constrain when they are applied to a simulation. Selectors can match well-known labels like `node.kubernetes.io/instance-type` or `karpenter.sh/node-pool` and custom labels defined in a NodePool's `.spec.labels` or `.spec.requirements`. When multiple NodeOverlays match an offering for a given simulation, they are merged together using `spec.weight` to resolve conflicts.
+
+### Examples
+
+Configure price to match a compute savings plan
+```yaml
+kind: NodeOverlay
+metadata:
+  name: default-discount
+spec:
+  pricePercent: 90
+```
+
+Support extended resource types (e.g. smarter-devices/fuse)
+* https://github.com/kubernetes-sigs/karpenter/issues/751
+* https://github.com/kubernetes-sigs/karpenter/issues/729
+
+```yaml
+kind: NodeOverlay
+metadata:
+  name: extended-resource
+spec:
+  selector:
+    matchLabels:
+      karpenter.sh/capacity-type: on-demand
+    matchExpressions:
+    - key: node.kubernetes.io/instance-type
+      operator: In
+      values: ["m5.large", "m5.2xlarge", "m5.4xlarge", "m5.8xlarge", "m5.12xlarge"]
+  capacity:
+    smarter-devices/fuse: 1
```

Comment:
To be clear, the "extended resource support" in this design would only support statically applied extended resources on selected nodes, correct? This would not cover cases where dynamic extended resources are needed, for example setting an extended resource for available network bandwidth which varies by EC2 instance type.

Additionally, karpenter still needs to have its provisioning and deprovisioning logic updated to account for arbitrary extended resources other than the gpu resource, right?

Reply:
> This would not cover cases where dynamic extended resources are needed, for example setting an extended resource for available network bandwidth which varies by EC2 instance type. 

You could achieve through multiple NodeOverlays, that match different selectors.

e.g. m5.large has 2, m5.2xlarge has 4.

Reply:
I don't think defining a NodeOverlay for each network bandwidth amount is a reasonable solution for my use case. We use a wide variety of instance types and sizes.

Do you think it could be feasible to incorporate some level of templating for values that can be easily retrieved from the EC2 API?

Reply:
For NetworkBandwidth, I'd love to just support this one first class. 

We could always try to do something that attempts to model this like: 
```
resourcesPerCPU: 
  network: 1Gi
```
But it's hard to capture all potential relationships without devolving into something like https://github.com/google/cel-spec

Reply:
Similar use case with KubeVirt, VM pods are getting "dynamic" label keys as node selectors, e.g. `scheduling.node.kubevirt.io/tsc-frequency-${cpuTimerValue}`

In my Karpenter fork, just providing an "ignored list" through environment variable where I'm doing pattern matching against `scheduling.node.kubevirt.io/tsc-frequency-*`